docs: analysis for non-k8s deployments (#2778)
keptn/lifecycle-toolkit: docs-new/docs/use-cases/non-k8s.md

# Keptn for non-Kubernetes deployments

Keptn can interact with deployments that are not running on Kubernetes
even though Keptn mainly targets Kubernetes deployments.
The following functionality is available for deployments not on Kubernetes:

- [Run Keptn tasks](#run-keptntask-for-a-deployment-not-on-kubernetes)
- [Run Keptn analysis](#run-keptn-analysis-for-a-deployment-not-on-kubernetes)

To use either of these features,
you must set up a Kubernetes cluster and
[install](../installation/index.md#basic-installation)
Keptn on it,
but this can be a very lightweight, single-node KinD cluster; see
[Create local Kubernetes cluster](../installation/k8s.md#create-local-kubernetes-cluster).
Keptn only runs  on-demand `KeptnTask` and `KeptnAnalysis` resources
so resource utilization is minimal.

## Run KeptnTask for a deployment not on Kubernetes

Keptn tasks running on a Kubernetes cluster can be triggered for
[workloads](https://kubernetes.io/docs/concepts/workloads/)
and applications that are deployed outside of Kubernetes.
For example, Keptn can run (or "trigger")
load and performance tests

> for an application that is deployed on a virtual machine.

Reviewer: ```suggestion
for an application that is deployed on a virtual machine,
or any other environment, as long as it can be reached
by the container running the KeptnTask.
```

kubectl apply -f test-task.yaml -n my-keptn-annotated-namespace
```

Applying this file causes Keptn to create a Kubernetes

> [job](https://kubernetes.io/docs/concepts/workloads/controllers/job/)

Reviewer: mentioning the Pod might not be needed here, as the creation of a Kubernetes Job implies that a Pod will then be created by the Job controller. But I also don't mind if we keep it to be more explicit

Author: I wondered about that myself but you had said it in your reply to Artur so I passed it through.  But I don't think mentioning the pod adds anything here so I deleted it.

kubectl get pods -n my-keptn-annotated-namespace
```

> See the

Reviewer: I think there is something missing, or this line can be removed

Author: I agree -- that is more confusing than helpful.  It's deleted

    workloadVersion: "1.0.1"
```

Use the following command to apply this resource:

```shell
kubectl apply -f test-task-2.yaml -n my-keptn-annotated-namespace
```

See the
[Deployment tasks](../guides/tasks.md)
guide and associated reference pages
for more information about running Keptn tasks
with deployments that do not run on Kubernetes.

## Run Keptn analysis for a deployment not on Kubernetes

The Keptn analyses feature
analyzes Service Level Objectives (SLOs)

> based on Service Level Inputs (SLIs).

Reviewer 2: ```suggestion
based on Service Level Indicators (SLIs).
```

It can apply weights to reach a composite score
about the health of the deployment,

> similar to what the Keptn v1 quality gates provided.

Reviewer 2: I would say similar to Keptn v1 evaluations

> similar to what the Keptn v1 quality gates provided.

Reviewer 2: ```suggestion
similar to what the Keptn v1 evaluations provided.
```

    workloadVersion: "1.0.1"
```

Use the following command to apply this resource:

```shell
kubectl apply -f test-task-2.yaml -n my-keptn-annotated-namespace
```

See the
[Deployment tasks](../guides/tasks.md)
guide and associated reference pages
for more information about running Keptn tasks
with deployments that do not run on Kubernetes.

## Run Keptn analysis for a deployment not on Kubernetes

The Keptn analyses feature
analyzes Service Level Objectives (SLOs)
based on Service Level Inputs (SLIs).
It can apply weights to reach a composite score
about the health of the deployment,
similar to what the Keptn v1 quality gates provided.
The data used can come from multiple instances
of multiple data providers
(such as Prometheus, Dynatrace, and DataDog).

A Keptn analysis can be run for any application running anywhere
as long Keptn can access a monitoring provider endpoint
that serves metrics for the application.
You can point to multiple instances of the supported monitoring providers
(Prometheus, Dynatrace, Datadog, and dql)

> so the application itself can run anywhere,

Reviewer: ```suggestion
so the application itself can run anywhere.
```

To implement a Keptn analysis for your deployment:

- Create a `KeptnMetricProvider` resource
  for each data source to be used for your analysis.
  This specifies the URL for the data source,
  assigns a `name` that Keptn uses to reference that provider,
  and can define a secret for the data provider if necessary.

- Create `ValueTemplate` resources for each SLI

>   and `AnalysisDefinition` resources for each SLO

Reviewer: ```suggestion
  and and `AnalysisDefinition` resource that contains all SLOs
```
